### PR TITLE
feat/AB#105843_EIS-nested-record-document-upload

### PR DESCRIPTION
**Pull request by unai-reliefapp**
# Description

feat: add nested file questions on creating new records upload option on parent form save

## Useful links

- [Link to ticket](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/105843)

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] Test A
- [ ] Test B

## Screenshots

Failed file upload is removed from the given record(default error handler)
https://github.com/user-attachments/assets/a7c1a23c-9f90-4a2e-9131-2f7f3941ff21

# Checklist:

( * == Mandatory ) 

- [X] * I have set myself as assignee of the pull request
- [X] * My code follows the style guidelines of this project
- [X] * Linting does not generate new warnings
- [X] * I have performed a self-review of my own code
- [X] * I have put the ticket for review, adding the oort-frontend team to the list of reviewers
- [X] * I have commented my code, particularly in hard-to-understand areas
- [X] * I have put JSDoc comment in all required places
- [X] * My changes generate no new warnings
- [X] * I have included screenshots describing my changes if relevant
- [X] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# More explanation
https://www.loom.com/share/05a716d61b9744faaf51fb304c21d1e5?sid=f87cf896-582a-4f76-93ae-8ceed801b145

### CODE REVIEW

**Comment by unai-reliefapp** [comment]
Hi @AntoineRelief , I have a first version here where we attach to the draft record question their temporary files, and once the user saves the parent form and the draft records are added, we upload their attached files.

Let me know if this is enough, and in error cases how you'll like to handle the form save flow

KR, Unai

**Comment by AntoineRelief** [comment]
@unai-reliefapp 
looks nice 👍 

good point for error handling, I'll need to play a bit with it to understand how we should do

**Comment by unai-reliefapp** [comment]
> @unai-reliefapp looks nice 👍
> 
> good point for error handling, I'll need to play a bit with it to understand how we should do

mostly here: https://github.com/ReliefApplications/ems-frontend/blob/232ae6350ec22da41949d868d57464973946b904/libs/shared/src/lib/services/form-helper/form-helper.service.ts#L349

Good thing of this approach is that we can nest records over records and still would try to upload attach files if exists 👍

File: libs/shared/src/lib/services/form-helper/form-helper.service.ts [hunk 1]
```diff
@@ -340,7 +344,19 @@ export class FormHelpersService {
                 data,
               },
             })
-          ).then(({ data }) => {
+          ).then(async ({ data }) => {
+            /** If any file attached to the temporary record, upload them once the record is created */
+            if (!isNil(element.question.temporaryFilesStorage)) {
+              try {
+                await this.uploadFiles(
+                  survey,
```
Comment (AntoineRelief) [inline libs/shared/src/lib/services/form-helper/form-helper.service.ts]: @unai-reliefapp 
in fact, I believe we should do the upload before the addition of a new record, and, if upload of a file fails, remove it from the record, and if possible, indicate that upload of nested file failed

Comment (unai-reliefapp) [reply libs/shared/src/lib/services/form-helper/form-helper.service.ts]: @AntoineRelief  Pushed suggested changes, the file upload error is handled as any other file upload, and the removal of failed file from question is extended as default behavior for all records.

Let me know if is it okay

KR, Unai

**Comment by AntoineRelief** [comment]
@unai-reliefapp 

so we will change how it works
everytime you add / edit a "nested" record, we will do the upload of files directly, and won't wait for the record to really be saved 
so, when we are closing the modal by clicking on save, for nested records, we should do the upload, and, if there is an issue, prevent the modal to be closed, and better mark the question with an error, if possible

**Comment by unai-reliefapp** [comment]
@AntoineRelief  pushed proposed changes plus fix the clear all files bug from the survey file question(it was not deleting any file from the temporaryFileStorage)
image

File: libs/shared/src/lib/components/form/form.component.ts [hunk 2]
```diff
@@ -270,7 +270,18 @@ export class FormComponent
         this.temporaryFilesStorage,
         this.form?.id
       );
-    } catch {
+    } catch (errors) {
+      /** If there is any upload errors, save them for display */
+      const uploadErrors = (errors as { question: string; file: File }[]).map(
```
Comment (AntoineRelief) [inline libs/shared/src/lib/components/form/form.component.ts:275]: @unai-reliefapp 
working fine! 
I also added the message there, so it's the same in all instances of the form

**Pull request merged**